engine/render/src/render/render.cpp

```diff
@@ -750,18 +750,36 @@ namespace dmRender
             return;
         }
 
+        Constant material_constant;
+        GetMaterialProgramConstant(material, name_hash, material_constant);
+
+        Vector4* value_ptr     = ro->m_ConstantsData;
+        Vector4* value_ptr_end = value_ptr + RenderObject::MAX_CONSTANT_COUNT;
+
         for (uint32_t i = 0; i < RenderObject::MAX_CONSTANT_COUNT; ++i)
         {
             Constant* c = &ro->m_Constants[i];
+
+            if ((value_ptr + num_values) > value_ptr_end)
+            {
+                dmLogError("Out of per object constant slots, max %d, when setting constant '%s' '", RenderObject::MAX_CONSTANT_COUNT, dmHashReverseSafe64(name_hash));
+                return;
+            }
+
             if (c->m_Location == -1 || c->m_NameHash == name_hash)
             {
                 // New or current slot found
-                c->m_Value = value;
+                c->m_ValuePtr      = value_ptr;
+                c->m_NumComponents = material_constant.m_NumComponents;
+                memcpy(c->m_ValuePtr, values, sizeof(values[0]) * num_values);
+
                 c->m_NameHash = name_hash;
-                c->m_Type = dmRenderDDF::MaterialDesc::CONSTANT_TYPE_USER;
+                c->m_Type     = num_values > 1 ? dmRenderDDF::MaterialDesc::CONSTANT_TYPE_USER_ARRAY : dmRenderDDF::MaterialDesc::CONSTANT_TYPE_USER;
```

Comment:
Is there a difference?
I mean, isn't it ok to use CONSTANT_TYPE_USER_ARRAY with a length of 1 ?

Reply:
No, not really. To be honest I'm not even sure we need the CONSTANT_TYPE_USER_ARRAY - the code paths are basically exactly the same..

Reply:
If we don't need it, I think it'd be good if we can avoid it.
Will the lack of it cause trouble in the editor, do you think?

Reply:
Removed